### PR TITLE
Check that the response isn't already closed

### LINKED ISSUE #255: IllegalStateException when deleting consumer while polling for records
Having an HTTP consumer polling for records and suddenly sending a DELETE request for removing itself from the bridge, can drive to the following error.

```
2019-06-17 11:36:16,695 [     vert.x-eventloop-thread-3] HttpBridgeContext              INFO  OpenAPI operation = poll
2019-06-17 11:36:16,695 [     vert.x-eventloop-thread-3] HttpSinkBridgeEndpoint         DEBUG [260617538] Request: body = null
2019-06-17 11:36:16,754 [     vert.x-eventloop-thread-3] HttpUtils                      INFO  [982326916] Request: method = DELETE, path = /consumers/my-group/instances/my-consumer
2019-06-17 11:36:16,754 [     vert.x-eventloop-thread-3] HttpUtils                      DEBUG [982326916] Request: headers = user-agent: Vert.x-WebClient/3.7.1
content-type: application/vnd.kafka.v2+json
host: new-host-3:8080

2019-06-17 11:36:16,754 [     vert.x-eventloop-thread-3] HttpBridgeContext              INFO  OpenAPI operation = deleteConsumer
2019-06-17 11:36:16,755 [     vert.x-eventloop-thread-3] HttpSinkBridgeEndpoint         DEBUG [982326916] Request: body = null
2019-06-17 11:36:16,756 [     vert.x-eventloop-thread-3] HttpSinkBridgeEndpoint         INFO  Deleted consumer my-consumer from group my-group
2019-06-17 11:36:16,756 [     vert.x-eventloop-thread-3] HttpUtils                      INFO  [982326916] Response: statusCode = 204, message = No Content
2019-06-17 11:36:16,756 [     vert.x-eventloop-thread-3] HttpUtils                      DEBUG [982326916] Response: headers = 
2019-06-17 11:36:16,757 [     vert.x-eventloop-thread-3] HttpUtils                      DEBUG [982326916] Response: body = []
2019-06-17 11:36:16,761 [vert.x-kafka-consumer-thread-0] AbstractCoordinator            INFO  [Consumer clientId=my-consumer, groupId=my-group] Sending LeaveGroup request to coordinator new-host-3.station:9092 (id: 2147483647 rack: null)
Jun 17, 2019 11:36:16 AM io.vertx.core.impl.ContextImpl
SEVERE: Unhandled exception
java.lang.IllegalStateException: Response has already been written
	at io.vertx.core.http.impl.HttpServerResponseImpl.checkValid(HttpServerResponseImpl.java:614)
HttpServerResponseImpl.java:614
	at io.vertx.core.http.impl.HttpServerResponseImpl.putHeader(HttpServerResponseImpl.java:193)
HttpServerResponseImpl.java:193
	at io.strimzi.kafka.bridge.http.HttpUtils.sendResponse(HttpUtils.java:23)
HttpUtils.java:23
	at io.strimzi.kafka.bridge.http.HttpSinkBridgeEndpoint.lambda$6(HttpSinkBridgeEndpoint.java:250)
HttpSinkBridgeEndpoint.java:250
	at io.strimzi.kafka.bridge.SinkBridgeEndpoint.handleManualRecords(SinkBridgeEndpoint.java:683)
SinkBridgeEndpoint.java:683
	at io.strimzi.kafka.bridge.SinkBridgeEndpoint.lambda$16(SinkBridgeEndpoint.java:705)
SinkBridgeEndpoint.java:705
	at io.vertx.kafka.client.consumer.impl.KafkaConsumerImpl.lambda$poll$16(KafkaConsumerImpl.java:560)
KafkaConsumerImpl.java:560
	at io.vertx.kafka.client.consumer.impl.KafkaReadStreamImpl.lambda$null$40(KafkaReadStreamImpl.java:731)
KafkaReadStreamImpl.java:731
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:320)
ContextImpl.java:320
	at io.vertx.core.impl.EventLoopContext.lambda$executeAsync$0(EventLoopContext.java:38)
EventLoopContext.java:38
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
AbstractEventExecutor.java:163
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404)
SingleThreadEventExecutor.java:404
```

After consumer deletion, the channel is closed and the bridge is trying to reply to the poll request.

### PR DESCRIPTION
**Pull request by ppatierno**
This PR fixes #255

### CODE REVIEW

File: src/main/java/io/strimzi/kafka/bridge/http/HttpUtils.java [hunk 1]
```diff
@@ -17,15 +17,19 @@
...
+            if (body != null) {
+                log.debug("[{}] Response: body = {}", routingContext.get("request-id"), Json.decodeValue(body));
+                routingContext.response().putHeader(HttpHeaderNames.CONTENT_TYPE, contentType);
+                routingContext.response().putHeader(HttpHeaderNames.CONTENT_LENGTH, String.valueOf(body.length()));
+                routingContext.response().write(body);
+            }
+            routingContext.response().end();
+            logResponse(routingContext);
+        } else {
+            log.warn("[{}] Response: already closed/ended!");
```
Comment (tombentley) [inline src/main/java/io/strimzi/kafka/bridge/http/HttpUtils.java]: Is this really deserving of warn? It seems to me that it would not be uncommon in normal (and correct) usage of the bridge to see this situation. And there's nothing the user of the bridge about it. In which case it should be at most info.

**Comment by ppatierno** [comment]
@tombentley that's a good catch! I did some tests and actually, the poll callback was using the "new" delete routing context instead of the "old/related" one.
Using in the callback a final variable assigned from the `routingContext` member instance seems to fix the issue, because the callback uses the right routing context.

**Comment by ppatierno** [comment]
Anyway, it's still worth checking that the routing context isn't closed or ended because after handling the delete and coming back to handle the right routing context for the last still ongoing poll, the right routing context is closed because the client closed the connection just after getting the response to the delete.
Maybe we could just avoid the warning log.

**Comment by ppatierno** [comment]
I left more comments and test cases on the related issue #255

File: src/main/java/io/strimzi/kafka/bridge/SinkBridgeEndpoint.java [hunk 2]
```diff
@@ -677,50 +673,38 @@ private void handleCommit(AsyncResult<Void> commitResult) {
...
-        this.commitOffsetsHandler = commitOffsetsHandler;
         this.consumer.commit(offsetsData, result -> {
-            if (this.commitOffsetsHandler != null) {
-                this.handleCommittedOffsets(result);
+            if (commitOffsetsHandler != null) {
+                if (result.succeeded()) {
+                    commitOffsetsHandler.handle(Future.succeededFuture(result.result()));
+                } else {
+                    commitOffsetsHandler.handle(Future.failedFuture(result.cause()));
+                }
```
[the remark this answers is not in the document]
Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/SinkBridgeEndpoint.java]: Yes and it should be even for the consume above other the other following commit as you already mentioned.

File: src/main/java/io/strimzi/kafka/bridge/http/HttpUtils.java [hunk 3]
```diff
@@ -17,15 +17,19 @@
...
+            if (body != null) {
+                log.debug("[{}] Response: body = {}", routingContext.get("request-id"), Json.decodeValue(body));
+                routingContext.response().putHeader(HttpHeaderNames.CONTENT_TYPE, contentType);
+                routingContext.response().putHeader(HttpHeaderNames.CONTENT_LENGTH, String.valueOf(body.length()));
+                routingContext.response().write(body);
+            }
+            routingContext.response().end();
+            logResponse(routingContext);
+        } else {
+            log.debug("[{}] Response: already closed/ended!", routingContext.get("request-id").toString());
```
[the remark this answers is not in the document]
Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/http/HttpUtils.java]: As explained in the related issue, the scenario is the following ...

If I kill the application I am using for tests, the shutdown hook undeploys the consumer verticle which stops the timer about sending poll requests and calls the delete consumer. At this point in the shutdown, I wait that the delete is done properly with a timeout: doing more long poll requests in the above scenarios it can happen that the delete is postponed so the application just ends for timeout on the latch. When on the bridge the poll will come back, it will try to write a response on a closed/ended socket. Because the consumer isn't deleted properly, the bridge has no knowledge that the application is no more interested in the poll responses.

File: src/main/java/io/strimzi/kafka/bridge/http/HttpSinkBridgeEndpoint.java [hunk 4]
```diff
@@ -127,9 +127,10 @@ private void doSeek(JsonObject bodyAsJson) {
             this.seek(topicPartition, offset, fut.completer());
         }
 
+        final RoutingContext rc = routingContext;
```
[the remark this answers is not in the document]
Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/http/HttpSinkBridgeEndpoint.java]: it's a fair point but the `handle` method comes from the `BridgeEndpoint` interface which is in common with AMQP bridging where the routing context has no meaning.

**Comment by ppatierno** [comment]
@tombentley it turned out that this PR mostly around the consume handler fixes another issue.
With the current bridge, when the topic doesn't exist and the cluster needs time to create it so warning like this are raised:

```
WARN  [Consumer clientId=my-consumer, groupId=my-group] Error while fetching metadata with correlation id 2 : {my-topic=LEADER_NOT_AVAILABLE}
```

when more poll requests comes into the bridge (i.e. every 1 sec) at some point this exception is raised as well:

```
Jun 18, 2019 12:26:48 PM io.vertx.core.impl.ContextImpl
SEVERE: Unhandled exception
java.lang.IllegalStateException: Response has already been written
	at io.vertx.core.http.impl.HttpServerResponseImpl.checkValid(HttpServerResponseImpl.java:614)
	at io.vertx.core.http.impl.HttpServerResponseImpl.putHeader(HttpServerResponseImpl.java:193)
	at io.strimzi.kafka.bridge.http.HttpUtils.sendResponse(HttpUtils.java:23)
	at io.strimzi.kafka.bridge.http.HttpSinkBridgeEndpoint.lambda$doPoll$5(HttpSinkBridgeEndpoint.java:250)
	at io.strimzi.kafka.bridge.SinkBridgeEndpoint.handleManualRecords(SinkBridgeEndpoint.java:683)
	at io.strimzi.kafka.bridge.SinkBridgeEndpoint.lambda$consume$10(SinkBridgeEndpoint.java:705)
	at io.vertx.kafka.client.consumer.impl.KafkaConsumerImpl.lambda$poll$16(KafkaConsumerImpl.java:560)
	at io.vertx.kafka.client.consumer.impl.KafkaReadStreamImpl.lambda$null$39(KafkaReadStreamImpl.java:729)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:320)
	at io.vertx.core.impl.EventLoopContext.lambda$executeAsync$0(EventLoopContext.java:38)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:495)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
```
wrong callback used with a routing context response already closed.
I tested the new bridge image and it doesn't happen anymore with the other 0.12.0 stuff from Strimzi operator.

File: src/main/java/io/strimzi/kafka/bridge/http/HttpUtils.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tombentley) [reply src/main/java/io/strimzi/kafka/bridge/http/HttpUtils.java]: I think you're conflating two different things in the code currently:

* The response is ended. It can only be the bridge which ended it. In which case trying to send a response to an already ended response must indicate a bug in the bridge and logging at debug is not the right thing to do.
* The TCP connection is closed. Presumably it could have been closed by the bridge, or on the client side.
    * If the connection was closed by the bridge (because the verticle is being undeployed, for example) then I would argue that it is again a bug in the bridge: when the bridge closes a TCP connection it should be waiting for outstanding responses to finish, or at least sending responses for any outstanding requests. However this is inherently racy (since the client can keep making requests while we're trying to send responses for existing outstanding requests), so maybe we need to think more about how to handle this corner case.
    * If the connection was closed by the client (which is, I think, the situation in your tests), then the debug logging stmt is fine, because there's nothing else the bridge can do.

File: src/main/java/io/strimzi/kafka/bridge/http/HttpSinkBridgeEndpoint.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (tombentley) [reply src/main/java/io/strimzi/kafka/bridge/http/HttpSinkBridgeEndpoint.java]: I'm not asking to change signature of the `handle()` method, just for it to get the RC and then pass it to the private methods in HttpSinkBridgeEndpoint, rather than using a field. Am I missing something obvious?

In any case, I think we've stretched the abstractions which try to paper over the differences between AMQP and HTTP beyond their useful limits. _Even if_ we keep support for AMQP, I suspect the protocol and API interactions are just too different to really be worth abstracting over.

File: src/main/java/io/strimzi/kafka/bridge/http/HttpUtils.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/http/HttpUtils.java]: > The response is ended. It can only be the bridge which ended it.

Is this assumption based on the fact that there is an `end()` method on the response. Maybe it's considered ended even when the remote client close the connection or maybe not. Btw the condition I have today is 
```
if (!routingContext.response().closed() && !routingContext.response().ended())
```

so maybe the trace I see is related to the fact that the response is closed and it happens even when remote client closes the connection.
Let me to be more precise about the conditions.

> If the connection was closed by the bridge (because the verticle is being undeployed, for example)

I am not closing the bridge, why it should close the connection to the client?

[the remark this answers is not in the document]
Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/http/HttpUtils.java]: Just adding that, using Wireshark, the FYN is sent by the client and that then the `Response: already closed/ended!` log is raised so I guess that it's because the response is closed not ended ... let me check ;)

[the remark this answers is not in the document]
Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/http/HttpUtils.java]: @tombentley we could have reached a good point where "no bugs" seem to be in the bridge. After my tests and logging response closed/ended more precisely ... I see that in my scenarios ...
```
2019-06-18 16:54:25,968 [     vert.x-eventloop-thread-3] HttpUtils                      DEBUG [1907003446] Response: already closed [true] - ended [false]
2019-06-18 16:54:30,969 [     vert.x-eventloop-thread-3] HttpUtils                      DEBUG [1036567481] Response: already closed [true] - ended [false]
2019-06-18 16:54:35,970 [     vert.x-eventloop-thread-3] HttpUtils                      DEBUG [500698412] Response: already closed [true] - ended [false]
2019-06-18 16:54:40,971 [     vert.x-eventloop-thread-3] HttpUtils                      DEBUG [1200619310] Response: already closed [true] - ended [false]
2019-06-18 16:54:45,972 [     vert.x-eventloop-thread-3] HttpUtils                      DEBUG [1652686025] Response: already closed [true] - ended [false]
```

The log is due to connection closed by the clients and never response ended as you mentioned by the bridge.
If it's fine with you now we should decide how to log this info. Anyway in this early age of the bridge having something logged could be really useful to study different corner cases ... maybe coming back to warn?

[the remark this answers is not in the document]
Comment (tombentley) [reply src/main/java/io/strimzi/kafka/bridge/http/HttpUtils.java]: > Is this assumption based on the fact that there is an end() method on the response.

That and the javadoc on `response.end()` and `response.ended()` makes it pretty clear to me that the ended state is exclusive of connection closed state.

>  Anyway in this early age of the bridge having something logged could be really useful to study different corner cases ... maybe coming back to warn?

How to you propose to get the logs for those different corner cases? If you can't get those logs then this extra logging is worthless. It would be better to log at WARN in those cases where you think it would really be a bug in the bridge. That way an absence of complaints about those warn messages is a little move convincing as evidence that the code is correct.

[the remark this answers is not in the document]
Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/http/HttpUtils.java]: For starting I would propose to log as WARN just when the response is ended which could be a bridge bug. Wdyt?

[the remark this answers is not in the document]
Comment (tombentley) [reply src/main/java/io/strimzi/kafka/bridge/http/HttpUtils.java]: That sounds good to me.

File: src/main/java/io/strimzi/kafka/bridge/http/HttpSinkBridgeEndpoint.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ppatierno) [reply src/main/java/io/strimzi/kafka/bridge/http/HttpSinkBridgeEndpoint.java]: @tombentley no you are right, I was thinking about that just now and I didn't read your comment. I will make this change as well before merging.

[the remark this answers is not in the document]
Comment (tombentley) [reply src/main/java/io/strimzi/kafka/bridge/http/HttpSinkBridgeEndpoint.java]: Thanks @ppatierno
